feat(alert, notice): Allow custom calcite-ui-icon overrides

macandcheese opened the pull request:
**Related Issue:** Resolves #1024 (cc @layesdesign @mitc7862 - I will open another issue for the different component request)
Resolves #1204 

## Summary
Add ability to override recommended status-based icon for alert, and notice
Adds examples to dev demo, storybook
Improve docs in input for overriding recommended type-based icon


<!--

Please make sure the PR title and/or commit message adheres to the https://conventionalcommits.org/ specification.

Note: If your PR only has one commit and it is NOT semantic, you will need to either

a. add another commit and wait for the check to update
b. proceed to squash merge, but make sure the commit message is the same as the title.

This is because of the way GitHub handles single-commit squash merges (see https://github.com/zeke/semantic-pull-requests/issues/17)

If this is component-related, please verify that:

- [ ] feature or fix has a corresponding test
- [ ] changes have been tested with demo page in Edge

-->

driskull commented on src/components/calcite-alert/calcite-alert.tsx:
```diff
@@ -79,6 +80,14 @@ export class CalciteAlert {
+    // if an icon string is not provided, but icon is true and a default icon is present
+    // for the requested color, set that as the icon
+    const colors = ["blue", "green", "red", "yellow"];
+    this.icon = this.icon
```
What if you have a stateful private property called `computedIcon` that watches `icon` and handles the boolean case?

Currently, if icon changes after initialization, the boolean case isn't being handled.

Thoughts?

macandcheese replied:
Yes, good catch. Will update on these new additions and the input component as well.

driskull replied:
Maybe move colors const to the resources file too. Could be used for testing

paulcpederson replied:
Would it not be fine to just do this in `renderIcon`?  Seems pretty lightweight to handle this logic during render...

driskull replied:
Yeah that's another option. Either way works since its lightweight

paulcpederson replied:
Agree that since alert and notice use the same logic it's a good idea to share in a resources file with some quick spec tests, that would be awesome

macandcheese replied:
Sounds good, I'll do some clean up on this.

driskull reviewed (commented):
comment

driskull commented on src/components/calcite-alert/calcite-alert.tsx:
```diff
@@ -259,10 +268,9 @@ export class CalciteAlert {
     return (
       <div class="alert-icon">
-        <calcite-icon icon={path} scale="m" />
+        <calcite-icon icon={this.icon as string} scale="m" />
```
the `as string` won't be necessary after using stateful prop

paulcpederson commented:
Once the `connectedCallback` stuff is moved, I think this is good to go, thanks @macandcheese

macandcheese commented:
Ok, I added a shared helper for the use case of "set a default icon from a predefined set when icon prop is requested as a boolean, or accept a custom icon string if requested" - and updated to use that on input, alert, and notice. 

Also does some general cleanup to use shared prop definitions, and moves the status color definitions and set of "status based icons" to a shared location. Give a 👀  and let me know!

paulcpederson commented on src/utils/dom.ts:
```diff
@@ -118,3 +118,16 @@ export function getDescribedByElement<T extends Element>(element: Element): T |
+  iconValue: string | boolean,
+  matchedValue: string
+): string {
+  if (iconValue && typeof iconValue === "string" && iconValue !== "") {
```
you may be able to skip the first one here. `typeof null` is `"null"` and `typeof undefined` is `"undefined"` so neither would break or be `"string"`, I think.

paulcpederson commented on src/components/calcite-alert/calcite-alert.tsx:
```diff
@@ -57,19 +59,24 @@ export class CalciteAlert {
+   * also pass a calcite-ui-icon name to this prop to display a requested icon */
+  @Prop({ reflect: true }) icon: string | boolean;
+
+  @Watch("icon") iconWatcher(): void {
```
Does this update the icon if I use `<calcite-alert icon color="blue">` and then change to `<calcite-alert icon color="green">`? I think you would need to watch both `icon` and `color` props. May be easier to just do this in the render like:

```
const icon = setRequestedIcon(StatusIcons, this.icon, this.color)
```

If you keep this as a prop, watchers have their own section in the file order, IIRC. and The requested icon may need to be a `@State` element to render after update, but I could be wrong about that?

driskull replied:
I think like this:

```ts
@State() requestedIcon: string;

@Watch("icon")
@Watch("color")
updateRequestedIcon(): string {
  this.requestedIcon = setRequestedIcon(StatusIcons, this.icon, this.color);
}

```

macandcheese replied:
I'll add a watch for color too. Not sure about file order - all the component files just list the watcher after the prop. Although if they are combined here I can just place at the end of the prop list... or create a // Watchers file section?

paulcpederson commented:
Wow, using `enum` and `Record` like a complete TypeScript boss. Had a couple comments, but looks good.

driskull commented on src/components/calcite-alert/calcite-alert.tsx:
```diff
@@ -190,49 +226,23 @@ export class CalciteAlert {
+    if (this.queue?.[0] === this.el) {
+      this.openAlert();
+      if (this.autoDismiss) {
+        setTimeout(() => this.closeAlert(), DURATIONS[this.autoDismissDuration]);
```
Is there any case where this timeout needs to be cleared?

macandcheese replied:
I don't believe so, this PR just moved the definitions to a resource file.

driskull commented on src/components/calcite-alert/calcite-alert.tsx:
```diff
@@ -248,22 +258,12 @@ export class CalciteAlert {
-    );
+  /** emit the opened alert and the queue */
+  private openAlert(): void {
+    setTimeout(() => (this.queued = false), 300);
```
Is there any case where this timeout needs to be cleared?

driskull reviewed (approved):
👍

macandcheese commented:
Merging 📁
